baseplate/config.py

```diff
@@ -333,30 +332,125 @@ def __init__(self):
         self.__dict__ = self
 
 
-def _parse_config_section(config, spec, root):
-    parsed = ConfigNamespace()
-    for key, parser_or_spec in spec.items():
-        assert "." not in key, "dots are not allowed in keys"
+class Parser(object):
+    """Base for config parsers."""
 
-        if root:
-            key_path = "%s.%s" % (root, key)
-        else:
-            key_path = key
-
-        if callable(parser_or_spec):
-            parser = parser_or_spec
-            raw_value = config.get(key_path, "")
-
-            try:
-                parsed[key] = parser(raw_value)
-            except Exception as e:
-                raise ConfigurationError(key, e)
-        elif isinstance(parser_or_spec, dict):
-            subspec = parser_or_spec
-            parsed[key] = _parse_config_section(config, subspec, root=key_path)
+    @classmethod
+    def from_spec(self, spec):
+        """Return a parser for the given spec object."""
+        if isinstance(spec, Parser):
+            return spec
+        elif isinstance(spec, dict):
+            return SpecParser(spec)
+        elif callable(spec):
+            return CallableParser(spec)
         else:
-            raise AssertionError("invalid specification: %r" % parser_or_spec)
-    return parsed
+            raise AssertionError("invalid specification: %r" % spec)
+
+    def parse(self, key_path, raw_config):
+        """Parse and return the relevant info for a given key.
+
+        :param str key_path: The key this parser is looking for.
+        :param dict raw_config: The full raw configuration dictionary.
+
+        """
+        raise NotImplementedError
+
+
+class SpecParser(Parser):
+    """A parser that validates a static specification."""
+
+    def __init__(self, spec):
+        self.spec = spec
+
+    def parse(self, root, raw_config):
+        parsed = ConfigNamespace()
+        for key, spec in self.spec.items():
+            assert "." not in key, "dots are not allowed in keys"
+
+            if root:
+                key_path = "%s.%s" % (root, key)
+            else:
+                key_path = key
+
+            parser = Parser.from_spec(spec)
+            parsed[key] = parser.parse(key_path, raw_config)
+        return parsed
+
+
+class CallableParser(Parser):
+    """A parser that wraps a simple callable."""
+
+    def __init__(self, callable):
+        self.callable = callable
+
+    def parse(self, key_path, raw_config):
+        raw_value = raw_config.get(key_path, "")
+
+        try:
+            return self.callable(raw_value)
+        except Exception as exc:
+            raise ConfigurationError(key_path, exc)
+
+
+class DictOf(Parser):
+    """A group of options of a given type.
```

Comment:
I guess I'll have to see how it comes out when the docs are generated, but I didn't really get what was going on here until I read `config_dictof_example.ini`.  Maybe it'll be obvious at that point though.